tunnistamo/tests/test_auth_error_redirect.py

```diff
@@ -128,3 +129,49 @@ def test_should_not_redirect_to_oidc_client_if_the_next_parameter_is_not_to_an_o
     # Should redirect to login
     assert response.status_code == 302
     assert response.url.startswith('/login/')
+
+
+@pytest.mark.django_db
+def test_on_auth_error_redirect_to_login_should_remember_idp_hint(
+    settings,
+    oidcclient_factory,
+    loginmethod_factory,
+):
+    settings.AUTHENTICATION_BACKENDS = settings.AUTHENTICATION_BACKENDS + (
+        'users.tests.conftest.DummyFixedOidcBackend',
+    )
+    settings.SOCIAL_AUTH_DUMMYFIXEDOIDCBACKEND_OIDC_ENDPOINT = 'https://dummy.example.com'
+    settings.SOCIAL_AUTH_DUMMYFIXEDOIDCBACKEND_ON_AUTH_ERROR_REDIRECT_TO_CLIENT = False
+    settings.SOCIAL_AUTH_FIELDS_STORED_IN_SESSION = ['idp_hint']
+    reload_social_django_utils()
+
+    django_client = CancelExampleComRedirectClient()
+
+    # Add multiple login methods to the client
+    login_methods = [
+        loginmethod_factory(provider_id='helsinki_adfs'),
+        loginmethod_factory(provider_id=DummyFixedOidcBackend.name),
+    ]
+
+    # Start the OIDC flow.
+    start_oidc_authorize(
+        django_client,
+        oidcclient_factory,
+        login_methods=login_methods,
+        # Set idp_hint to make Tunnistamo redirect the client to the
+        # DummyFixedOidcBackend
+        extra_authorize_params={'idp_hint': DummyFixedOidcBackend.name}
+    )
+
+    # Return the user from the social auth with an error.
+    response = _request_social_auth_complete_with_error(django_client)
+
+    assert response.status_code == 302
+    assert response.url.startswith('/login/')
+
+    # Validate that the idp_hint parameter is still present when the user is redirected
+    # back to the login view after unsuccessful social login
+    response_url_parts = list(urlparse(response.url))
+    response_query_parameters = QueryDict(response_url_parts[4])
```

Comment:
This part could be a bit more readable without the magic index 4. Perhaps something like:
```suggestion
    response_url_parts = urlparse(response.url)
    response_query_parameters = QueryDict(response_url_parts.query)
```